agent/pool/pool.go

```diff
@@ -629,9 +610,20 @@ func (p *ConnPool) rpc(dc string, nodeName string, addr net.Addr, method string,
 		return fmt.Errorf("rpc error getting client: %w", err)
 	}
 
-	// Use the zero value if the request doesn't implement RPCInfo
-	if info, ok := args.(structs.RPCInfo); ok {
-		sc.stream.FirstReadTimeout = info.Timeout(p.Timeout, p.MaxQueryTime, p.DefaultQueryTime)
+	var deadline time.Time
+	if info, ok := args.(BlockableQuery); ok {
+		// Timeout here is calculated differently based on blocking vs non-blocking query.
+		timeout := info.BlockingTimeout(p.MaxQueryTime, p.DefaultQueryTime)
+		if timeout <= 0 {
+			// must be non-blocking
+			timeout = p.ReadTimeout
+		}
+		if timeout > 0 {
+			deadline = time.Now().Add(timeout)
+		}
+	}
+	if err := sc.stream.SetReadDeadline(deadline); err != nil {
```

Comment:
It looks like `deadline` isn't initialized if we cannot cast to a BlockableQuery. Will that cause issues for this `SetReadDeadline(deadline)` that will receive a timestamp from the past?

Reply:
A zero value for `SetReadDeadline` will set no deadline, which is intended for writes and I think an acceptable fallback for any other types of queries that don't implement `BlockableQuery`

Reply:
I'm having a hard time working out whether this is the same behaviour as before - did `RPCInfo` only get implemented for Reads? If not then it seems like this new code is now never setting a timeout on write requests where it was before.

Separate issue: assuming it's the same, is there a reason we allow writes to take arbitrarily long? I could understand having a separate timeout for them because they might have to wait for Raft etc. but it seems like it would be ideal if we didn't have zero timeout on writes...

That said, (assuming the behaviour is same as before) it's better to do that in another PR.

Reply:
Hmm I took a look at the original PR and it seems to apply `rpc_hold_timeout` to all RPCs (I conflated `net.Conn.Read` with "read" requests). 

I think the correct solution would be to rename `rpc_read_timeout` to `rpc_client_timeout` and have it apply the same read timeout to all requests (queries, writes, basically anything except blocking-queries).

Reply:
@banks I understand the concern about writes. But given the fact that we had 7s timeouts on writes in the last few releases and the only issues raised by the community and customers were regarding non-blocking queries, I think it's safe to apply a larger (maybe 300s) timeout across the board. What do you think?

Reply:
We can solve the issue of not having timeouts on writes first and if anything comes up where we find that write RPCs need to be tuned separately, maybe another config may be more appropriate.

Reply:
Oh nice. Yeah I couldn't recall if this was everything. If the old behaviour was to timeout writes too then I'd def not want to _remove_ that!

> I think the correct solution would be to rename rpc_read_timeout to rpc_client_timeout and have it apply the same read timeout to all requests (queries, writes, basically anything except blocking-queries).

Yeah this sounds ideal to me for now. We can add more config if we really need to.